File: src/Atlasd/Battlenet/Protocols/Game/ChatCommands/AwayCommand.cs
using Atlasd.Localization;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Atlasd.Battlenet.Protocols.Game.ChatCommands
{
    class AwayCommand : ChatCommand
    {
        public AwayCommand(byte[] rawBuffer, List<string> arguments) : base(rawBuffer, arguments) { }

        public override bool CanInvoke(ChatCommandContext context)
        {
            return context != null && context.GameState != null && context.GameState.ActiveAccount != null;
        }

        public override void Invoke(ChatCommandContext context)
        {
            string message = Arguments.Count == 0 ? null : string.Join(" ", Arguments);
            string r; // reply

            if (context.GameState.Away == null || !string.IsNullOrEmpty(message))
            {
                context.GameState.Away = string.IsNullOrEmpty(message) ? "Not available" : message;
                r = Resources.AwayCommandOn;
            } else
            {
                context.GameState.Away = null;
                r = Resources.AwayCommandOff;
            }

            foreach (var kv in context.Environment)
            {
                r = r.Replace("{" + kv.Key + "}", kv.Value);
            }

            foreach (var line in r.Split(Battlenet.Common.NewLine))
            {
                new ChatEvent(ChatEvent.EventIds.EID_INFO, context.GameState.ChannelFlags, context.GameState.Client.RemoteIPAddress, context.GameState.Ping, context.GameState.OnlineName, line).WriteTo(context.GameState.Client);
            }
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/Game/ChatCommands/AwayCommand.vb
Imports AtlasV.Localization
Imports System
Imports System.Collections.Generic
Imports System.Linq

Namespace AtlasV.Battlenet.Protocols.Game.ChatCommands
    Class AwayCommand
        Inherits ChatCommand

        Public Sub New(ByVal varRawBuffer As Byte(), ByVal varArguments As List(Of String))
            MyBase.New(varRawBuffer, varArguments)
        End Sub

        Public Overrides Function CanInvoke(ByVal varContext As ChatCommandContext) As Boolean
            Return varContext IsNot Nothing AndAlso varContext.GameState IsNot Nothing AndAlso varContext.GameState.ActiveAccount IsNot Nothing
        End Function

        Public Overrides Sub Invoke(ByVal varContext As ChatCommandContext)
            Dim message As String = If(Arguments.Count = 0, Nothing, String.Join(" ", Arguments))
            Dim r As String

            If varContext.GameState.Away Is Nothing OrElse Not String.IsNullOrEmpty(message) Then
                varContext.GameState.Away = If(String.IsNullOrEmpty(message), "Not available", message)
                r = Resources.AwayCommandOn
            Else
                varContext.GameState.Away = Nothing
                r = Resources.AwayCommandOff
            End If

            For Each kv In varContext.Environment
                r = r.Replace("{" & kv.Key & "}", kv.Value)
            Next

            For Each line In r.Split(Battlenet.Common.NewLine)
                Call New ChatEvent(ChatEvent.EventIds.EID_INFO, varContext.GameState.ChannelFlags, varContext.GameState.Ping, varContext.GameState.OnlineName, line).WriteTo(varContext.GameState.Client)
            Next
        End Sub
    End Class
End Namespace
